INTERNALS.md

```diff
@@ -0,0 +1,226 @@
+For example, we don't want to mark an image for deletion the second the last server group using it has been terminated. 
+Within a reasonable time frame we might want to roll back to that image.
+
+One example of that is the [ResourceUseTrackingRepository](swabbie-core/src/main/kotlin/com/netflix/spinnaker/swabbie/repository/ResourceUseTrackingRepository.kt), which is used to record use and check if a resource has been used or not.
+
+
+### Exclusion Policies
+
+Resources can be excluded/opted out from consideration using exclusion policies.
+
+- [ResourceExclusionPolicy](swabbie-core/src/main/kotlin/com/netflix/spinnaker/swabbie/exclusions/ExclusionPolicy.kt): Excludes resources at runtime
+
+
+### Allowlisting
+
+Allowlisting is part of the exclusion mechanism. When defined, only resources allowlisted will be considered, skipping everything else not allowed.
+Typically we create an allowlist on the `owner` key so that we can roll out deletion of a resource to our team before the rest of the company.
+
+### Opting Out
+
+Any resource can be opted out of deletion, which means it will never be deleted.
+We use redis to keep track of this opted out state. 
+Since redis _is not_ a durable data store, we also recommend implementing a cloud provider specific tagging for all resources.
+For aws, we tag each opted out resource with `expiration_time: never`.
+
+Any resource tagged with `expiration_time: never` will never be deleted. 
+
+### Deleting
+
+We use orca and clouddriver to delete resources.
+When we submit resources to delete to orca we use an ad-hoc pipeline with a wait stage (waiting a random amount of time) before the delete.
+This helps us not exceed rate limits for deletion.
+
+### Storage: Redis
+
+Swabbie uses redis to keep track of locking and to store information about a resource.
+
+Here's an example of the locking options:
+```
+locking:
+  enabled: true
+  maximumLockDurationMillis: 360000
+  heartbeatRateMillis: 5000
+  leaseDurationMillis: 30000
+```
+
+Scheduling the cleanup of resources is done by keeping an index of visited resources in a `ZSET`, using the projected deletion time as the `score`.
+Getting elements from the `ZSET` from `-inf` to `now` will return all resources ready to be deleted.
+Getting elements from the `ZSET` from `0.0` to `+inf` will return all currently marked resources.
+
+
+### Notifications
+
+When resources are marked for deletion, a notification is sent to the owner.
+Resource owners are resolved using resolution strategies. Default strategy is getting the email field on the resource's application.
+The deletion of a resource will be adjusted when the notification is sent, respecting the retention days for the resource type.
+
+### Dry Run
+
+This will ensure swabbie runs in dryRun, meaning no writes, nor any destructive action of the data will occur
+```
+swabbie:
+  dryRun: true
+```
+It's also possible to turn on dryRun at a resource type level
+
+### Configuration
```

Comment:
I don't think `Configuration` belongs in the `INTERNALS.md` -- this should continue to live in `README.md` imo.

Reply:
Ok, I'll move it.